worker/externalcontrollerupdater/externalcontrollerupdater.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package externalcontrollerupdater
+
+import (
+	"io"
+	"reflect"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/crosscontroller"
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+var logger = loggo.GetLogger("juju.worker.externalcontrollerupdater")
+
+type ExternalControllerUpdaterClient interface {
```

Comment:
Would it be better not to group these 3 methods on a single interface? For a given external controller, we need to watch for changes and then ask for changes when the watcher fires. And then we need to take those changes and update the local controller. So Watch/GetInfo belong on one interface, and SetInfo() on another interface.

Reply:
What does that buy? In practice they're all implemented by the same API, and conceptually they belong together as they're all dealing with the same source of information. i.e. we're watching, fetching and updating the local controller's knowledge of external controllers.

Reply:
The same API but different targets. Just as Go uses small interfaces specific to a particular purpose, and then composes as needed into broader interfaces, I think we should do that here too (expect there's no need for the broader interface, just the individual ones). The facade connected to the local controller will just implement SetExternalControllerInfo(), and the one connected to the remote controller will implement the other two.

Reply:
No, that's not how it works.

All of the methods in this interface are dealing with the local controller.
 - WatchExternalControllers watches for addition and removal of external controller records in the local controller
 - ExternalControllerInfo asks the local controller for its current information about an external controller (so we can make an API connection)
 - ExternalControllerInfo sets new external controller information in the local controller

The `ExternalControllerWatcherClient` interface a couple of types below is the one that communicates with the external controller. It has two methods:
 - WatchControllerInfo watches the external controller's API info for changes
 - ControllerInfo gets the external controller's current API info